--- nodogsplash-0.9_beta9.9.1/src/firewall.c.orig	2009-01-24 21:32:11.000000000 +0100
+++ nodogsplash-0.9_beta9.9.1/src/firewall.c	2009-01-24 21:32:22.000000000 +0100
@@ -95,7 +95,7 @@
 
   /* Find ip, copy mac in reply */
   reply = NULL;
-  while (!feof(proc) && (fscanf(proc, " %15[0-9.] %*s %*s %17[A-F0-9:] %*s %*s", ip, mac) == 2)) {
+  while (!feof(proc) && (fscanf(proc, " %15[0-9.] %*s %*s %17[a-fA-F0-9:] %*s %*s", ip, mac) == 2)) {
     if (strcmp(ip, req_ip) == 0) {
       reply = safe_strdup(mac);
       break;
