--- nds/src/fw_iptables.c.orig	2009-11-16 21:44:08.000000000 +0100
+++ nds/src/fw_iptables.c	2009-11-16 21:53:39.000000000 +0100
@@ -154,6 +154,7 @@
   char *fmt_cmd,
     *cmd;
   int rc;
+  int i;
 
   va_start(vlist, format);
   safe_vasprintf(&fmt_cmd, format, vlist);
@@ -164,8 +165,18 @@
   free(fmt_cmd);
 
   debug(LOG_DEBUG, "Executing command: %s", cmd);
-	
-  rc = execute(cmd, fw_quiet);
+  
+  for (i = 0; i < 5; i++) {
+    rc = execute(cmd, fw_quiet);
+    if (rc == 4) {
+      /* iptables error code 4 indicates a resource problem that might
+       * be temporary. So we retry to insert the rule a couple of times. */
+      sleep(1);
+    } else {
+      break;
+    }
+  }
+
   if(!fw_quiet && rc != 0) {
     debug(LOG_ERR, "Nonzero exit status %d from command: %s", rc, cmd);
   }
