--- busybox-1.4.2.orig/networking/httpd.c	2008-04-11 17:34:12.000000000 +0200
+++ busybox-1.4.2/networking/httpd.c	2008-04-11 18:51:33.000000000 +0200
@@ -858,6 +858,9 @@
 		numNames = sizeof(httpResponseNames) / sizeof(httpResponseNames[0])
 	};
 
+	if ((config->redirectPath) && (responseNum == HTTP_NOT_FOUND))
+		responseNum = HTTP_MOVED_TEMPORARILY;
+
 	for (i = 0; i < numNames; i++) {
 		if (httpResponseNames[i].type == responseNum) {
 			responseString = httpResponseNames[i].name;
@@ -886,7 +889,7 @@
 		len += sprintf(buf+len, "Location: %s%s%s%s%s%s\r\n",
 				(config->redirectHost ? "http://" : ""),
 				(config->redirectHost ? config->redirectHost : ""),
-				config->found_moved_temporarily,
+				(config->found_moved_temporarily ? config->found_moved_temporarily : config->redirectPath),
 				(config->redirectHost ? "" : "/"),
 				(config->query ? "?" : ""),
 				(config->query ? config->query : ""));
@@ -1603,10 +1606,6 @@
 		*++purl = '\0';       /* so keep last character */
 		test = purl;          /* end ptr */
 
-		/* redirect active */
-		if (config->redirectPath && (strncmp(url, config->redirectPath, strlen(config->redirectPath)) != 0))
-			config->found_moved_temporarily = config->redirectPath;
-
 		/* If URL is directory, adding '/' */
 		if(!config->redirectPath && (test[-1] != '/')) {
 			if (is_directory(url + 1, 1, &sb)) {
